services/m3dbnode/server/server_test.go

```diff
@@ -0,0 +1,476 @@
+		SetHeartbeatInterval(10 * time.Second)
+
+	err = svcs.SetMetadata(serviceID, metadata)
+	require.NoError(t, err)
+
+	placementOpts := placement.NewOptions().
+		SetValidZone(serviceZone)
+	placementSvc, err := svcs.PlacementService(serviceID, placementOpts)
+	require.NoError(t, err)
+
+	instance := placement.NewInstance().
+		SetID(hostID).
+		SetEndpoint(endpoint("127.0.0.1", servicePort)).
+		SetPort(servicePort).
+		SetRack("local").
+		SetWeight(1).
+		SetZone(serviceZone)
+	instances := []placement.Instance{instance}
+	shards := 256
+	replicas := 1
+	_, err = placementSvc.BuildInitialPlacement(instances, shards, replicas)
+	require.NoError(t, err)
+
+	// Setup the namespace
+	ns, err := newNamespaceProtoValue(namespaceID)
+	require.NoError(t, err)
+
+	kvStore, err := configSvcClient.KV()
+	require.NoError(t, err)
+
+	_, err = kvStore.Set(kvconfig.NamespacesKey, ns)
+	require.NoError(t, err)
+
+	// Run server
+	var (
+		interruptCh = make(chan error, 1)
+		bootstrapCh = make(chan struct{}, 1)
+		serverWg    sync.WaitGroup
+	)
+	serverWg.Add(1)
+	go func() {
+		Run(RunOptions{
+			ConfigFile:  configFd.Name(),
+			BootstrapCh: bootstrapCh,
+			InterruptCh: interruptCh,
+		})
+		serverWg.Done()
+	}()
+
+	// Wait for bootstrap
+	<-bootstrapCh
+
+	// Create client, read and write some data
+	// NB(r): Make sure client config points to the root config
+	// service since we're going to instantiate the client configuration
+	// just by itself.
+	cfg.Client.ConfigService = &cfg.ConfigService
+
+	cli, err := cfg.Client.NewClient(client.ConfigurationParameters{})
+	require.NoError(t, err)
+
+	session, err := cli.DefaultSession()
+	require.NoError(t, err)
+
+	defer session.Close()
+
+	start := time.Now().Add(-time.Minute)
+	values := []struct {
+		value float64
+		at    time.Time
+		unit  xtime.Unit
+	}{
+		{value: 1.0, at: start, unit: xtime.Second},
+		{value: 2.0, at: start.Add(1 * time.Second), unit: xtime.Second},
+		{value: 3.0, at: start.Add(2 * time.Second), unit: xtime.Second},
+	}
+
+	for _, v := range values {
+		err := session.Write(namespaceID, "foo", v.at, v.value, v.unit, nil)
+		require.NoError(t, err)
+	}
+
+	// Account for first value inserted at xtime.Second precision
+	fetchStart := start.Truncate(time.Second)
+
+	// Account for last value being inserted at xtime.Second and
+	// the "end" param to fetch being exclusive
+	fetchEnd := values[len(values)-1].at.Truncate(time.Second).Add(time.Nanosecond)
+
+	iter, err := session.Fetch(namespaceID, "foo", fetchStart, fetchEnd)
+	require.NoError(t, err)
+
+	for _, v := range values {
+		require.True(t, iter.Next())
+		dp, unit, _ := iter.Current()
+		assert.Equal(t, v.value, dp.Value)
+		// Account for xtime.Second precision on values going in
+		expectAt := v.at.Truncate(time.Second)
+		assert.Equal(t, expectAt, dp.Timestamp)
+		assert.Equal(t, v.unit, unit)
+	}
+
+	// Wait for server to stop
+	interruptCh <- fmt.Errorf("test complete")
+	serverWg.Wait()
+}
+
+var testConfig = `
```

Comment:
this is dope af! 

how would you feel about making this an exported type in a non `_test.go` file in the same package?

Reply:
i.e. some kind of method that takes options and spits out a YAML?
```go
type ConfigYAMLOptions struct {
  ServicePort int
  //...
}
func ConfigYAML(opts ConfigYAMLOptions) {
  //...
}
```

The only issue with this is that if you truly want to be able to print any type of YAML (i.e. make it flexible) then you need `ConfigYAMLOptions` to be as expressive as `config.Configuration` which parses the YAML to be.  As such it would be better to actually create a `config.Configuration` then marshal it to YAML for such a purpose.

What's the intended use case?  Then we can understand the best ergonomics to meet that use case.

Reply:
Yep so pretty much the use-case you're thinking: use this template, and write a tool in the m3db repo which populates it with the values to spit out a YAML. I hear your point about it needing to be as expressive as the struct itself in the general case. IMO the base version you have here would go a long way to get people started. Consider the template we use internally, most of it is static too.

Don't have to write the tool now, I figure just exporting the struct is a good enough starting point.